### ISSUE TITLE
Unique names for PulseAudio sink inputs

### ISSUE THREAD

**Issue by LiteracyFanatic** [qutebrowser/qutebrowser#5914]
After we have #5673, another nice enhancement would be to use the PID to match PulseAudio sink inputs and add the corresponding tab name to the properties so they can be told apart in `pactl` and `pavucontrol`.

**Comment by The-Compiler** [comment]
This isn't possible to implement - setting the PulseAudio name is done by the underlying Chromium/QtWebEngine by default, and can only be overridden by qutebrowser globally, via an environment variable (see #3832).

Related Chromium issues were closed as well I'm afraid:

- [768136 - Feature request: Tab title in PulseAudio stream name - chromium](https://bugs.chromium.org/p/chromium/issues/detail?id=768136&q=pulseaudio%20sink%20name&can=1)
- [847589 - set pulseaudio context name per web app, not as Chrome/Chromium - chromium](https://bugs.chromium.org/p/chromium/issues/detail?id=847589)

**Comment by LiteracyFanatic** [comment]
Wouldn't this be possible at runtime after the creation of the sink using `pacmd update-sink-proplist` (or whatever the programatic API call equivalent is) so long as we can identify the sinks by PID?

**Comment by The-Compiler** [comment]
Sinks and sources are outputs/inputs - I assume you mean `update-sink-input-proplist`, correct? I tried with `pacmd update-sink-input-proplist 221 media.name="blabla"` and that indeed seems to work.

The main problem I can see with that: qutebrowser has no idea when the underlying audio streams are created. The best it can get is checking when a tab plays audio and then add some kind of cool-down timer, but it all seems quite hacky. Additionally, a tab's renderer process can change when a different website is visited...

There *might* be a way to get around all this (perhaps there's some API to let Pulseaudio notify us when there's a new stream?), but it's certainly all a bit more hacky than I'd like. Reopening for now, though.

**Comment by LiteracyFanatic** [comment]
Yeah, whatever command actually works. I always get confused by the terminology. :sweat_smile:

Fortunately PulseAudio does have an API to monitor stream creation and modification. It can be monitored from the command line with `pactl subscribe` or in Python using the `pulsectl` package (I've got a script that shows a graphical volume bar using it). I assume the underlying mechanism is a D-BUS API we could call directly too.

**Comment by andrewcarlotti** [comment]
Using process ids is not necessarily sufficient (with current process creation), since it is possible to have multiple sink-inputs from the same process id (e.g. open youtube.com, Ctrl+click on the logo to open a second youtube.com tab, and then choose a video to play in each). The only distinction in that case appears to be the sequential client number that pulseaudio allocates, but I don't know whether that's accessible from qutebrowser or even from the QtWebEngineProcess.

However, you could still narrow it down a bit using process ids, and if tab<->process allocation changes in the future then labelling PulseAudio sink inputs might become more feasible. (Incidentally, I'm still haven't worked out how the process allocation works at the moment - it seems that old processes can be reused for new independent tabs, but I haven't investigated why/when this happens).

**Comment by LiteracyFanatic** [comment]
Indeed that does complicate things. 🤔
